### PR TITLE
Master website chatter jem

### PR DESCRIPTION
**Pull request by jem-odoo**
**Task :** https://www.odoo.com/web#id=30059&view_type=form&model=project.task&action=333&active_id=809&menu_id=4720

**Purpose :** 
In all website modules (blog, ecommerce, ...) we can post comments. But having a huge number of comments can be problemtic, since they all will be displayed on the same page (loading page time can be .... long).
The udemy comment zone is quite smart (see https://www.udemy.com/learn-html5-programming-from-scratch/ for instance). There is a pager for comments, and you can filter the rating (ecommerce support rating).

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_mail/controllers/main.py [hunk 1]
```diff
@@ -126,25 +133,52 @@ def is_follower(self, model, res_id, **post):
...
         return request.redirect(url)
+
+    def _website_message_post(self, res_model, res_id, message, **kw):
+        if not message:
+            return False
+        return _message_post_helper(res_model, int(res_id), message, **kw)
+
+    @http.route('/website_mail/init', type='json', auth='public', website=True)
+    def website_chatter_init(self, res_model, res_id, domain=False, limit=False, **kwargs):
+        is_user_public = bool(request.env.user == request.website.user_id)
```
Comment (tde-banana-odoo) [inline addons/website_mail/controllers/main.py]: equal already returns a bool you trou du cul

File: addons/website_mail/controllers/main.py [hunk 2]
```diff
@@ -126,25 +133,52 @@ def is_follower(self, model, res_id, **post):
...
+                'partner_id': request.env.user.partner_id.id
+            }
+        }
+
+    @http.route('/website_mail/fetch', type='json', auth='public', website=True)
+    def website_message_fetch(self, res_model, res_id, domain=False, limit=False, offset=False, **kw):
+        if not limit:
+            limit = 10
+        if not offset:
+            offset = 0
```
Comment (tde-banana-odoo) [inline addons/website_mail/controllers/main.py]: why are these not default parameters?

File: addons/website_mail/controllers/main.py [hunk 3]
```diff
@@ -126,25 +133,52 @@ def is_follower(self, model, res_id, **post):
...
+        if not domain:
+            domain = []
+        # Only search into website_message_ids, so apply the same domain to perform only one search
+        domain += request.env[res_model]._default_domain_website_message_ids()
+        domain += [('res_id', '=', res_id)]
```
Comment (tde-banana-odoo) [inline addons/website_mail/controllers/main.py]: why you no append?

File: addons/website_mail/static/src/js/website_mail.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+                'pager_scope': 5,
+                'pager_start': 1,
+                'is_user_public': true,
+                'is_user_publisher': false,
+                'domain': [],
+            }, options || {});
+            this.set('messages', []);
+            this.set('message_count', this.options['message_count']);
+            this.set('pager', {});
+            this.set('domain', this.options['domain']);
```
Comment (tde-banana-odoo) [inline addons/website_mail/static/src/js/website_mail.js]: I think you can pass object/dict to Widget#set

File: addons/website_mail/static/src/js/website_mail.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+                this._load_templates()
+            ).then(function(result){
+                // bind events
+                self.on("change:messages", self, self.render_messages);
+                self.on("change:message_count", self, function(){
+                    self.render_message_count();
+                    self.set('pager', self._pager(self._current_page));
+                });
+                self.on("change:pager", self, self.render_pager);
+                self.on("change:domain", self, self.on_change_domain);
```
Comment (tde-banana-odoo) [inline addons/website_mail/static/src/js/website_mail.js]: Why not set events unconditionally or via the class parameter thing? Do you really need to wait for templates to be loaded to hook these?

File: addons/website_mail/static/src/js/website_mail.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+            if(pmax - pmin < scope){
+                if(pmax - scope > 0){
+                    pmin = pmax - scope;
+                }else{
+                    pmin = 1;
+                }
+            }
+            var pages = [];
+            _.each(_.range(pmin, pmax+1), function(index){
```
Comment (tde-banana-odoo) [inline addons/website_mail/static/src/js/website_mail.js]: Why not just _.range(0, pmax+1 - pmin)?

File: addons/website_rating/models/mail_message.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+            message.rating_value = mapping.get(message.id, 0.0)
+
+    def _search_rating_value(self, operator, value):
+        self.env.cr.execute("""
+            SELECT rating, message_id
+            FROM rating_rating
+            WHERE rating = %s AND consumed='t' AND message_id IS NOT NULL
+            ORDER BY create_date DESC""", (value,)
+        )
+        message_ids = [m['message_id'] for m in self.env.cr.dictfetchall()]
```
Comment (tde-banana-odoo) [inline addons/website_rating/models/mail_message.py]: why select rating if not using it? And why not use fetchall and get relevant field by index?

File: addons/website_rating/models/mail_message.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+class MailMessage(models.Model):
+    _inherit = 'mail.message'
+
+    rating_value = fields.Float("Rating Value", compute='_compute_rating_value', search='_search_rating_value')
+
+    @api.multi
+    def _compute_rating_value(self):
+        ratings = self.env['rating.rating'].search([('message_id', 'in', self.ids), ('consumed', '=', True)], order='create_date DESC')
+        mapping = dict((r.message_id.id, r.rating) for r in ratings)
```
Comment (tde-banana-odoo) [inline addons/website_rating/models/mail_message.py]: Use dictionary comprehension

File: addons/website_rating/models/mail_message.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        message_ids = [m['message_id'] for m in self.env.cr.dictfetchall()]
+        return [('id', 'in', message_ids)]
+
+    @api.multi
+    def website_message_format(self):
+        result = super(MailMessage, self).website_message_format()
+        # inlude rating value in data if requested
+        if self._context.get('rating_include'):
+            message_data = self.read(['rating_value'])
+            message_tree = dict((m['id'], m['rating_value']) for m in message_data)
```
Comment (tde-banana-odoo) [inline addons/website_rating/models/mail_message.py]: Use dictionary comprehension, and is there reason to use read instead of regular browsing?

File: addons/website_mail/controllers/main.py [hunk 10]
```diff
@@ -1,12 +1,18 @@
...
+
 from werkzeug.exceptions import NotFound
 
 from odoo import http
 from odoo.http import request
 
+
+def _special_access_object(res_model, res_id, token='', token_field=''):
+    record = request.env[res_model].browse(res_id)
+    if token and record and token == getattr(record.sudo(), token_field, None):
```
Comment (tde-banana-odoo) [inline addons/website_mail/controllers/main.py]: Shouldn't we use consteq to compare tokens ?

File: addons/website_mail/controllers/main.py [hunk 11]
```diff
@@ -27,7 +33,9 @@ def _message_post_helper(res_model='', res_id=None, message='', token='', token_
     """
     record = request.env[res_model].browse(res_id)
     author_id = request.env.user.partner_id.id if request.env.user.partner_id else False
-    if token and record and token == getattr(record.sudo(), token_field, None):
+
+    access_as_sudo = _special_access_object(res_model, res_id, token=token, token_field=token_field)
```
Comment (tde-banana-odoo) [inline addons/website_mail/controllers/main.py]: Shouldn't we do something if the token is not correct ? If I understand well the record will not be accessed as sudo, but maybe it could just avoid doing stuff if a wrong token is given.

File: addons/website_mail/controllers/main.py [hunk 12]
```diff
@@ -97,25 +105,56 @@ def is_follower(self, model, res_id, **post):
...
+    def chatter_post(self, res_model, res_id, message, **kw):
         url = request.httprequest.referrer
         if message:
-            message = _message_post_helper(res_model, int(res_id), message, **kw)
-            url = url + "#message-%s" % (message.id,)
+            self._website_message_post(res_model, res_id, message, **kw)
+            url = url + "#discussion"
         return request.redirect(url)
+
+    def _website_message_post(self, res_model, res_id, message, **kw):
```
Comment (tde-banana-odoo) [inline addons/website_mail/controllers/main.py]: What is the purpose of that wrapper ?

File: addons/website_mail/controllers/main.py [hunk 13]
```diff
@@ -97,25 +105,56 @@ def is_follower(self, model, res_id, **post):
...
-            'body': msg.body,
-            'date': msg.date,
-            'author': msg.author_id.name,
-            'image_url': '/mail/%s/%s/avatar/%s' % (msg.model, msg.res_id, msg.author_id.id)
-        }
-        return data
-
     @http.route(['/website_mail/post/post'], type='http', methods=['POST'], auth='public', website=True)
-    def chatter_post(self, res_model='', res_id=None, message='', redirect=None, **kw):
+    def chatter_post(self, res_model, res_id, message, **kw):
```
Comment (tde-banana-odoo) [inline addons/website_mail/controllers/main.py]: Just a naming thing: chatter_post, website_message_post, message_post_helper, website_chatter_init, website_message_fetch -> there is some space for prefixing convention cleaning ;) .

File: addons/website_mail/controllers/main.py [hunk 14]
```diff
@@ -97,25 +105,56 @@ def is_follower(self, model, res_id, **post):
...
+                'is_user_public': is_user_public,
+                'is_user_publisher': request.env.user.has_group('website.group_website_publisher'),
+                'display_composer': display_composer,
+                'partner_id': request.env.user.partner_id.id
+            }
+        }
+
+    @http.route('/website_mail/fetch', type='json', auth='public', website=True)
+    def website_message_fetch(self, res_model, res_id, domain=False, limit=False, offset=False, **kw):
+        if not limit:
```
Comment (tde-banana-odoo) [inline addons/website_mail/controllers/main.py]: Why not putting it directly in the parameters ?

File: addons/website_mail/controllers/main.py [hunk 15]
```diff
@@ -97,25 +105,56 @@ def is_follower(self, model, res_id, **post):
...
+            limit = 10
+        if not offset:
+            offset = 0
+        if not domain:
+            domain = []
+        # Only search into website_message_ids, so apply the same domain to perform only one search
+        # extract domain from the 'website_message_ids' field
+        field_domain = request.env[res_model]._fields['website_message_ids'].domain
+        domain += field_domain(request.env[res_model].browse(res_id)) if callable(field_domain) else field_domain
```
Comment (tde-banana-odoo) [inline addons/website_mail/controllers/main.py]: Actually I don't think we have to browse res_id. Indeed domains are only model dependent, not record dependent at the ORM level. I think we should keep the same semantic.

File: addons/website_mail/controllers/main.py [hunk 16]
```diff
@@ -97,25 +105,56 @@ def is_follower(self, model, res_id, **post):
...
+        domain += [('res_id', '=', res_id)]
+        # Check access
+        Message = request.env['mail.message']
+        access_as_sudo = _special_access_object(res_model, res_id, token=kw.get('token'), token_field=kw.get('token_field'))
```
Comment (tde-banana-odoo) [inline addons/website_mail/controllers/main.py]: For example here, shouldn't a wrong token send you to hell, instead of probably crashing due to access rights issues ?

File: addons/website_mail/security/website_mail_security.xml [hunk 17]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (tde-banana-odoo) [inline addons/website_mail/security/website_mail_security.xml]: Void file ?

File: addons/website_mail/static/src/js/website_mail.js [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+     */
+    var WebsiteMailThread = Widget.extend({
+        template: 'website_mail.chatter',
+        events: {
+            "click .o_website_mail_pager_btn": 'on_click_pager'
+        },
+
+        init: function(parent, options){
+            this._super.apply(this, arguments);
+            console.log($(parent).data());
```
Comment (tde-banana-odoo) [inline addons/website_mail/static/src/js/website_mail.js]: logger :)

File: addons/website_mail/static/src/js/website_mail.js [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+                self.set('message_count', self.options['message_count']);
+                self.set('messages', self.preprocess_messages(result['messages']));
+            });
+        },
+        _load_templates: function(){
+            return ajax.loadXML('/website_mail/static/src/xml/website_mail.xml', qweb);
+        },
+        fetch_message: function(domain){
+            var self = this;
+            var data = this._prepare_fetch_message_params();
```
Comment (tde-banana-odoo) [inline addons/website_mail/static/src/js/website_mail.js]: Unnecessary variable

File: addons/website_mail/static/src/js/website_mail.js [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+                // set options and parameters
+                self.options = _.extend(self.options, result['options'] || {});
+                self.set('message_count', self.options['message_count']);
+                self.set('messages', self.preprocess_messages(result['messages']));
+            });
+        },
+        _load_templates: function(){
+            return ajax.loadXML('/website_mail/static/src/xml/website_mail.xml', qweb);
+        },
+        fetch_message: function(domain){
```
Comment (tde-banana-odoo) [inline addons/website_mail/static/src/js/website_mail.js]: Shouldn't we use the same naming convention, and use message_fetch, message_fetch_prepare, ... ?

File: addons/website_mail/static/src/js/website_mail.js [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+     *  - Display chatter: pager, total message, composer (according to access right)
+     *  - Provider API to filter displayed messages
+     */
+    var WebsiteMailThread = Widget.extend({
+        template: 'website_mail.chatter',
+        events: {
+            "click .o_website_mail_pager_btn": 'on_click_pager'
+        },
+
+        init: function(parent, options){
```
Comment (tde-banana-odoo) [inline addons/website_mail/static/src/js/website_mail.js]: There are some leftover loggers to remove .

File: addons/website_mail/static/src/js/website_mail.js [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+            this._super.apply(this, arguments);
+            console.log($(parent).data());
+            this.options = _.defaults($(parent).data(), {
```
Comment (tde-banana-odoo) [inline addons/website_mail/static/src/js/website_mail.js]: Strange construct, where does it come from ?

File: addons/website_mail/static/src/js/website_mail.js [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+            });
+        },
+        preprocess_messages: function(messages){
+            _.each(messages, function(m){
+                m['author_avatar_url'] = _.str.sprintf('/web/image/%s/%s/author_avatar/50x50', 'mail.message', m.id);
+                m['published_date_str'] = _.str.sprintf(_t('Published on %s'), moment(m.date).format('MMMM Do YYYY, h:mm:ss a'));
+            });
+            return messages;
+        },
+        _pager: function(page, total, step, scope){
```
Comment (tde-banana-odoo) [inline addons/website_mail/static/src/js/website_mail.js]: I would either remove parameters or explicitely set them to this.get(parameter) when calling the function, to lessen the number of variable processing.

File: addons/website_mail/static/src/js/website_mail.js [hunk 24]
```diff
@@ -0,0 +1,193 @@
...
+            // Compute Pager
+            var page_count = parseInt(Math.ceil(parseFloat(total) / step));
+
+            var page = Math.max(1, Math.min(parseInt(page), page_count));
+            scope -= 1;
+
+            var pmin = Math.max(page - parseInt(Math.floor(scope/2)), 1);
+            var pmax = Math.min(pmin + scope, page_count);
+
+            if(pmax - pmin < scope){
```
Comment (tde-banana-odoo) [inline addons/website_mail/static/src/js/website_mail.js]: Whut ?

File: addons/website_mail/static/src/js/website_mail.js [hunk 25]
```diff
@@ -0,0 +1,193 @@
...
+    var _t = core._t;
+
+    /**
+     * Widget WebsiteMailThread
+     *
+     *  - Fetch message fron controller
+     *  - Display chatter: pager, total message, composer (according to access right)
+     *  - Provider API to filter displayed messages
+     */
+    var WebsiteMailThread = Widget.extend({
```
Comment (tde-banana-odoo) [inline addons/website_mail/static/src/js/website_mail.js]: General note: some space issues (missing spaces, ...), a linting is probably necessary.

File: addons/website_mail/static/src/xml/website_mail.xml [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+
+    <t t-name="website_mail.chatter_message_count">
+        <t t-set="count" t-value="widget.get('message_count')"/>
+        <div class="o_message_counter">
+            <t t-if="count">
+                <span class="o_message_count fa fa-comments"> <t t-esc="count"/></span>
+                <t t-if="count == 1">comment</t>
+                <t t-if="count != 1">comments</t>
+            </t>
+            <t t-if="!count">
```
Comment (tde-banana-odoo) [inline addons/website_mail/static/src/xml/website_mail.xml]: t-else ?

File: addons/website_mail/controllers/main.py [hunk 27]
```diff
@@ -149,6 +149,9 @@ def website_message_fetch(self, res_model, res_id, domain=False, limit=False, of
         field_domain = request.env[res_model]._fields['website_message_ids'].domain
         domain += field_domain(request.env[res_model].browse(res_id)) if callable(field_domain) else field_domain
         domain += [('res_id', '=', res_id)]
+        # None published can see only the published messages
```
Comment (tde-banana-odoo) [inline addons/website_mail/controllers/main.py]: Should probably be in its own commit.

File: addons/website_mail/static/src/js/website_mail.js [hunk 28]
```diff
@@ -37,38 +37,33 @@ odoo.define('website_mail.thread', function(require) {
                 'is_user_publisher': false,
                 'domain': [],
             }, options || {});
-            console.log('options=', this.options);
             this.set('messages', []);
             this.set('message_count', this.options['message_count']);
-            this.set('current_page', this.options['pager_start']);
             this.set('pager', {});
             this.set('domain', this.options['domain']);
+            this._current_page = this.options['pager_start'];
```
Comment (tde-banana-odoo) [inline addons/website_mail/static/src/js/website_mail.js]: Shouldn't those changes be in the commit about new front-end widget ?

File: addons/website_rating/__manifest__.py [hunk 29]
```diff
@@ -0,0 +1,17 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+{
+    'name': 'Website Rating',
+    'category': 'Hidden',
+    'version': '1.0',
+    'description': """
+Bridge module for rating on website stuffs.
```
Comment (tde-banana-odoo) [inline addons/website_rating/__manifest__.py]: website stuffs, especially with an 's', is quite strange.

File: addons/website_rating/__manifest__.py [hunk 30]
```diff
@@ -0,0 +1,17 @@
...
+        """,
+    'depends': ['rating', 'website_mail'],
+    'data': [
+        'views/website_rating_templates.xml',
+    ],
+    'qweb': ['static/src/xml/*.xml'],
+    'auto_installable': True,
```
Comment (tde-banana-odoo) [inline addons/website_rating/__manifest__.py]: Isn't it 'auto_install' ?

File: addons/website_rating/controllers/website_mail.py [hunk 31]
```diff
@@ -0,0 +1,49 @@
...
+
+
+class WebsiteRating(WebsiteMail):
+
+    def _website_message_post(self, res_model, res_id, message, **kw):
+        rating_value = kw.pop('rating_value', None)
+        result = super(WebsiteRating, self)._website_message_post(res_model, res_id, message, **kw)
+
+        # if message is posted, and rating value given, then create a consumed rating
+        if not isinstance(result, NotFound):
```
Comment (tde-banana-odoo) [inline addons/website_rating/controllers/website_mail.py]: Concatenate ifs .

File: addons/website_rating/controllers/website_mail.py [hunk 32]
```diff
@@ -0,0 +1,49 @@
...
+
+from odoo import http
+from odoo.http import request
+
+from odoo.addons.website_mail.controllers.main import WebsiteMail
+
+
+class WebsiteRating(WebsiteMail):
+
+    def _website_message_post(self, res_model, res_id, message, **kw):
```
Comment (tde-banana-odoo) [inline addons/website_rating/controllers/website_mail.py]: It seems strange this behavior is implemented here. Shouldn't it be implemented at model level, somewhere like in an override of mail.thread in rating ?

File: addons/website_rating/models/mail_message.py [hunk 33]
```diff
@@ -0,0 +1,38 @@
...
+from odoo import api, fields, models
+
+
+class MailMessage(models.Model):
+    _inherit = 'mail.message'
+
+    rating_value = fields.Float("Rating Value", compute='_compute_rating_value', search='_search_rating_value')
+
+    @api.multi
+    def _compute_rating_value(self):
```
Comment (tde-banana-odoo) [inline addons/website_rating/models/mail_message.py]: Is the rating value the last given value, am I right ? A small comment could help here I think.

File: addons/website_rating/models/mail_message.py [hunk 34]
```diff
@@ -0,0 +1,38 @@
...
+        ratings = self.env['rating.rating'].search([('message_id', 'in', self.ids), ('consumed', '=', True)], order='create_date DESC')
+        mapping = dict((r.message_id.id, r.rating) for r in ratings)
+        for message in self:
+            message.rating_value = mapping.get(message.id, 0.0)
+
+    def _search_rating_value(self, operator, value):
+        self.env.cr.execute("""
```
Comment (tde-banana-odoo) [inline addons/website_rating/models/mail_message.py]: I fear it will become quickly hard to perform a search. Once you have a lot of ratings, you will have a domain term like [('id', 'in', [A MAGNIFICIENT ARRAY])]. Don't you think we should store the rating value in order to search efficiently on it ?

File: addons/website_rating/static/src/js/website_mail.js [hunk 35]
```diff
@@ -0,0 +1,186 @@
...
+        onchange_rating_domain_value: function(){
+            var domain = [];
+            if(this.get('rating_value')){
+                domain = [['rating_value', '=', this.get('rating_value')]];
+            }
+            console.log('change to ',this.get('rating_value'));
+            this.change_current_page(1, domain);
+        },
+        // utils
+        round_to_half: function(value) {
```
Comment (tde-banana-odoo) [inline addons/website_rating/static/src/js/website_mail.js]: I am not a JS expert, but don't we have access to some method allowing to do it :p ?

File: addons/website_rating/static/src/js/website_mail.js [hunk 36]
```diff
@@ -0,0 +1,186 @@
...
+            this.$('.o_website_rating_select').css({
+                'opacity': 1,
+            });
+        },
+        onchange_rating_domain_value: function(){
+            var domain = [];
+            if(this.get('rating_value')){
+                domain = [['rating_value', '=', this.get('rating_value')]];
+            }
+            console.log('change to ',this.get('rating_value'));
```
Comment (tde-banana-odoo) [inline addons/website_rating/static/src/js/website_mail.js]: Leftover log.

File: addons/website_sale/models/__init__.py [hunk 37]
```diff
@@ -1,5 +1,4 @@
 import ir_http
-import rating
```
Comment (tde-banana-odoo) [inline addons/website_sale/models/__init__.py]: Mmmh ?

File: addons/website_mail/security/website_mail_security.xml [hunk 38]
```diff
@@ -2,6 +2,12 @@
 <odoo>
     <data noupdate="1">
 
+        <record id="mail_message_rule_public" model="ir.rule">
```
Comment (tde-banana-odoo) [inline addons/website_mail/security/website_mail_security.xml]: There is some mismatch in the commits. One commit add a hardcoded domain, one commit adds a void file, and this one finishes the work by removing the hardcoded domain and adding stuff in the void file. I fear you will have to do strange stuff when rebasing. Bon courage.

File: addons/website_mail/controllers/main.py (hunk 10 above)
Comment (jem-odoo) [inline addons/website_mail/controllers/main.py]: Ok, but it have to check than the token_field is filled to avoid given None as argument to consteq. So the condition looks like : 
`if token and record and getattr(record.sudo(), token_field, None) and consteq(getattr(record.sudo(), token_field, None), str(token)):`

File: addons/website_mail/controllers/main.py (hunk 11 above)
Comment (jem-odoo) [inline addons/website_mail/controllers/main.py]: Like raising a `forbidden` ? It could be cool indeed !

File: addons/website_mail/controllers/main.py (hunk 12 above)
Comment (jem-odoo) [inline addons/website_mail/controllers/main.py]: It is a hook, to create the rating in website_rating. Otherwise, super will return the url, and don't have the message_id to associate the rating with.
Maybe `_message_post_helper` should be a controller method. What do you think ?

File: addons/website_mail/controllers/main.py (hunk 15 above)
Comment (jem-odoo) [inline addons/website_mail/controllers/main.py]: Smart guy !

File: addons/website_mail/static/src/xml/website_mail.xml (hunk 26 above)
Comment (jem-odoo) [inline addons/website_mail/static/src/xml/website_mail.xml]: I didn't it was supported in qweb js. Great news !

File: addons/website_rating/__manifest__.py (hunk 30 above)
Comment (jem-odoo) [inline addons/website_rating/__manifest__.py]: :scream:

Comment (tde-banana-odoo) [inline addons/website_rating/__manifest__.py]: :dancing_women:

File: addons/website_mail/static/src/xml/website_mail.xml (hunk 26 above)
Comment (tde-banana-odoo) [inline addons/website_mail/static/src/xml/website_mail.xml]: Actually, it might not be. I don't know the status of js qweb. I forgot that subtle distinction.

File: addons/website_mail/security/website_mail_security.xml (hunk 17 above)
Comment (jem-odoo) [inline addons/website_mail/security/website_mail_security.xml]: Parce que c'est smart. Didjou !

File: addons/website_sale/models/__init__.py (hunk 37 above)
Comment (jem-odoo) [inline addons/website_sale/models/__init__.py]: Normal, since I have moved this file from website_sale ;)

File: addons/website_mail/static/src/js/website_mail.js (hunk 24 above)
Comment (jem-odoo) [inline addons/website_mail/static/src/js/website_mail.js]: I simply transform the algo from python side. Without trying to understand ;)

File: addons/website_mail/static/src/xml/website_mail.xml [hunk 39]
```diff
@@ -0,0 +1,132 @@
...
+                                <textarea rows="4" name="message" class="form-control" placeholder="Write a message..."></textarea>
+                                <input type='hidden' name="res_model" t-att-value="widget.options['res_model']"/>
+                                <input type='hidden' name="res_id" t-att-value="widget.options['res_id']"/>
+                                <input type='hidden' name="token" t-att-value="widget.options['token']" t-if="widget.options['token']"/>
+                                <input type='hidden' name="sha_in" t-att-value="widget.options['sha_in']" t-if="widget.options['sha_in']"/>
+                                <input type='hidden' name="sha_time" t-att-value="widget.options['sha_time']" t-if="widget.options['sha_time']"/>
+                                <input type='hidden' name="token_field" t-att-value="widget.options['token_field']" t-if="widget.options['token_field']"/>
+                                <div class="alert alert-danger mt8 mb0 o_website_mail_composer_error" style="display:none;">
+                                    Oops! Something went wrong. Try to reload the page and to log in.
```
Comment (Icallhimtest) [inline addons/website_mail/static/src/xml/website_mail.xml]: gotto make a choice "and log in" or "to log in" but not "and to log in" ^^

File: addons/website_rating/static/src/js/website_mail.js [hunk 40]
```diff
@@ -0,0 +1,233 @@
...
+        },
+
+        //--------------------------------------------------------------------------
+        // Public
+        //--------------------------------------------------------------------------
+
+        /**
+         * Update the messages format
+         *
+         * @params {Array<Object>}
```
Comment (Icallhimtest) [inline addons/website_rating/static/src/js/website_mail.js]: `@param`

comment to propagate in this file too

File: addons/website_rating/static/src/js/website_mail.js [hunk 41]
```diff
@@ -0,0 +1,233 @@
...
+            var decimal = (converted - parseInt(converted, 10));
+            decimal = Math.round(decimal * 10);
+            if(decimal == 5){
+                return (parseInt(converted, 10)+0.5);
+            }
+            if((decimal < 3) || (decimal > 7)){
+                return Math.round(converted);
+            }else{
+                return (parseInt(converted, 10)+0.5);
+            }
```
Comment (Icallhimtest) [inline addons/website_rating/static/src/js/website_mail.js]: I'd use Math.round(number * 2) / 2

File: addons/website_rating/static/src/js/website_mail.js [hunk 42]
```diff
@@ -0,0 +1,233 @@
...
+            this.$('.rate_text .label').text(this.labels[index]);
+        },
+        _onMouseleaveStar: function(e){
+            this.$('.rate_text').hide();
+        },
+        /**
+         * @private
+         * @param {MouseEvent} event
+         */
+        _onMoveStar: function(e){
```
Comment (Icallhimtest) [inline addons/website_rating/static/src/js/website_mail.js]: while testing the stars on the website, I found it a bit odd that after you click a certain rating, if you happen to hover on another star, your clicked rating gets lost.

Also, if you send a rating without any message, you get a refresh, but the rating doesn't go through. It'd be better if it didn't submit but had something like when you don't fill a required field on a form.

File: addons/website_mail/views/website_mail_templates.xml [hunk 43]
```diff
@@ -20,77 +20,26 @@
...
-                                    </div>
-                            </div>
-                            </div>
-                        </li>
-                    </t>
-                </t>
-            </ul>
-        </section>
+        <div id="discussion" class="hidden-print o_website_mail_thread"
+            t-att-data-token="token" t-att-public-user="website.env.user == request.website.user_id" t-att-data-res_model="object._name" t-att-data-res_id="object.id" t-att-data-token_field="token_field" t-att-data-pager_step="message_per_page or 10" t-att-data-allow_composer="'0' if disable_composer else '1'">
```
Comment (jem-odoo) [inline addons/website_mail/views/website_mail_templates.xml]: Setting  `t-att-token=="object[token_field]"` will crash if token_field is not set. And since we usually use the template without the token mecanism, it should be better like this ;)

Comment (Icallhimtest) [inline addons/website_mail/views/website_mail_templates.xml]: You're right, it'd be object.get(token_field) ^^. It's just a redundant argument imo, but I won't tilt if you wanna keep it this way.

File: addons/website_mail/static/src/js/website_mail.js [hunk 44]
```diff
@@ -0,0 +1,241 @@
...
+        //--------------------------------------------------------------------------
+
+        /**
+         * Change the current page by refreshing current domain
+         *
+         * @private
+         * @param {Number} page
+         * @param {Array} domain
+         */
+        _changeCurrentPage: function(page, domain){
```
Comment (jem-odoo) [inline addons/website_mail/static/src/js/website_mail.js]: Yes that for this, but when changing domain, I have to reset the `current_page` to trigger the messages fetch. But if I set the current_page to 1, and I'm already on the page one, replacing `1` by `1` does not trigger the fetching. So I have to use the domain, which is an object to trigger the fetching.

File: addons/website_rating/static/src/js/website_mail.js (hunk 42 above)
Comment (jem-odoo) [inline addons/website_rating/static/src/js/website_mail.js]: and it's done !

File: addons/website_mail/static/src/js/website_mail.js [hunk 45]
```diff
@@ -0,0 +1,241 @@
...
+            this._current_page = page;
+            var d = _.clone(this.get('domain'))
+            if(domain){
+                d = domain;
+            }
+            this.set('domain', d); // trigger fetch message
```
Comment (Icallhimtest) [inline addons/website_mail/static/src/js/website_mail.js]: if we keep the onchange triggers this way: `var d = (domain ? domain : _.clone(this.get('domain'))`

But imo changing the domain should change the page (or force its onchange if no page change), and changing the page should trigger the message fetching.
